error: path parameters cannot be ignored
  --> $DIR/typed-uris-bad-params.rs:79:37
   |
79 |     uri!(optionals: id = 10, name = _);
   |                                     ^

error: path parameters cannot be ignored
  --> $DIR/typed-uris-bad-params.rs:76:26
   |
76 |     uri!(optionals: id = _, name = "bob".into());
   |                          ^

error: invalid parameters for `has_two` route uri
  --> $DIR/typed-uris-bad-params.rs:72:19
   |
72 |     uri!(has_two: id = 100, cookies = "hi"); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32, name: String
   = help: missing parameter: `name`
help: unknown parameter: `cookies`
  --> $DIR/typed-uris-bad-params.rs:72:29
   |
72 |     uri!(has_two: id = 100, cookies = "hi"); //~ ERROR invalid parameters
   |                             ^^^^^^^

error: invalid parameters for `has_two` route uri
  --> $DIR/typed-uris-bad-params.rs:67:19
   |
67 |     uri!(has_two: cookies = "hi", id = 100, id = 10, id = 10); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32, name: String
   = help: missing parameter: `name`
help: unknown parameter: `cookies`
  --> $DIR/typed-uris-bad-params.rs:67:19
   |
67 |     uri!(has_two: cookies = "hi", id = 100, id = 10, id = 10); //~ ERROR invalid parameters
   |                   ^^^^^^^
help: duplicate parameter: `id`
  --> $DIR/typed-uris-bad-params.rs:67:45
   |
67 |     uri!(has_two: cookies = "hi", id = 100, id = 10, id = 10); //~ ERROR invalid parameters
   |                                             ^^       ^^

error: invalid parameters for `has_two` route uri
  --> $DIR/typed-uris-bad-params.rs:64:19
   |
64 |     uri!(has_two: name = "hi"); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32, name: String
   = help: missing parameter: `id`

error: invalid parameters for `has_two` route uri
  --> $DIR/typed-uris-bad-params.rs:60:19
   |
60 |     uri!(has_two: id = 100, id = 100, ); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32, name: String
   = help: missing parameter: `name`
help: duplicate parameter: `id`
  --> $DIR/typed-uris-bad-params.rs:60:29
   |
60 |     uri!(has_two: id = 100, id = 100, ); //~ ERROR invalid parameters
   |                             ^^

error: invalid parameters for `has_one_guarded` route uri
  --> $DIR/typed-uris-bad-params.rs:57:27
   |
57 |     uri!(has_one_guarded: id = 100, cookies = "hi"); //~ ERROR invalid parameters
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: unknown parameter: `cookies`
  --> $DIR/typed-uris-bad-params.rs:57:37
   |
57 |     uri!(has_one_guarded: id = 100, cookies = "hi"); //~ ERROR invalid parameters
   |                                     ^^^^^^^

error: invalid parameters for `has_one_guarded` route uri
  --> $DIR/typed-uris-bad-params.rs:54:27
   |
54 |     uri!(has_one_guarded: cookies = "hi", id = 100); //~ ERROR invalid parameters
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: unknown parameter: `cookies`
  --> $DIR/typed-uris-bad-params.rs:54:27
   |
54 |     uri!(has_one_guarded: cookies = "hi", id = 100); //~ ERROR invalid parameters
   |                           ^^^^^^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:50:19
   |
50 |     uri!(has_one: name = "hi"); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
   = help: missing parameter: `id`
help: unknown parameter: `name`
  --> $DIR/typed-uris-bad-params.rs:50:19
   |
50 |     uri!(has_one: name = "hi"); //~ ERROR invalid parameters
   |                   ^^^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:47:19
   |
47 |     uri!(has_one: id = 100, id = 100, ); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: duplicate parameter: `id`
  --> $DIR/typed-uris-bad-params.rs:47:29
   |
47 |     uri!(has_one: id = 100, id = 100, ); //~ ERROR invalid parameters
   |                             ^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:44:19
   |
44 |     uri!(has_one: id = 100, id = 100); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: duplicate parameter: `id`
  --> $DIR/typed-uris-bad-params.rs:44:29
   |
44 |     uri!(has_one: id = 100, id = 100); //~ ERROR invalid parameters
   |                             ^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:40:19
   |
40 |     uri!(has_one: name = 100, age = 50, id = 100, id = 50); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: unknown parameters: `name`, `age`
  --> $DIR/typed-uris-bad-params.rs:40:19
   |
40 |     uri!(has_one: name = 100, age = 50, id = 100, id = 50); //~ ERROR invalid parameters
   |                   ^^^^        ^^^
help: duplicate parameter: `id`
  --> $DIR/typed-uris-bad-params.rs:40:51
   |
40 |     uri!(has_one: name = 100, age = 50, id = 100, id = 50); //~ ERROR invalid parameters
   |                                                   ^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:37:19
   |
37 |     uri!(has_one: name = 100, age = 50, id = 100); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: unknown parameters: `name`, `age`
  --> $DIR/typed-uris-bad-params.rs:37:19
   |
37 |     uri!(has_one: name = 100, age = 50, id = 100); //~ ERROR invalid parameters
   |                   ^^^^        ^^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:34:19
   |
34 |     uri!(has_one: name = 100, id = 100); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: unknown parameter: `name`
  --> $DIR/typed-uris-bad-params.rs:34:19
   |
34 |     uri!(has_one: name = 100, id = 100); //~ ERROR invalid parameters
   |                   ^^^^

error: invalid parameters for `has_one` route uri
  --> $DIR/typed-uris-bad-params.rs:31:19
   |
31 |     uri!(has_one: id = 100, name = "hi"); //~ ERROR invalid parameters
   |                   ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: uri parameters are: id: i32
help: unknown parameter: `name`
  --> $DIR/typed-uris-bad-params.rs:31:29
   |
31 |     uri!(has_one: id = 100, name = "hi"); //~ ERROR invalid parameters
   |                             ^^^^

error: `has_two` route uri expects 2 parameters but 1 was supplied
  --> $DIR/typed-uris-bad-params.rs:29:19
   |
29 |     uri!(has_two: 10); //~ ERROR expects 2 parameters but 1
   |                   ^^
   |
   = note: expected parameters: id: i32, name: String

error: `has_two` route uri expects 2 parameters but 3 were supplied
  --> $DIR/typed-uris-bad-params.rs:28:19
   |
28 |     uri!(has_two: 10, "hi", "there"); //~ ERROR expects 2 parameters but 3
   |                   ^^^^^^^^^^^^^^^^^
   |
   = note: expected parameters: id: i32, name: String

error: `has_one_guarded` route uri expects 1 parameter but 2 were supplied
  --> $DIR/typed-uris-bad-params.rs:26:27
   |
26 |     uri!(has_one_guarded: "hi", 100); //~ ERROR expects 1 parameter but 2
   |                           ^^^^^^^^^
   |
   = note: expected parameter: id: i32

error: `has_one` route uri expects 1 parameter but 2 were supplied
  --> $DIR/typed-uris-bad-params.rs:25:19
   |
25 |     uri!(has_one: "Hello", 23, ); //~ ERROR expects 1 parameter but 2
   |                   ^^^^^^^^^^^^
   |
   = note: expected parameter: id: i32

error: `has_one` route uri expects 1 parameter but 2 were supplied
  --> $DIR/typed-uris-bad-params.rs:24:19
   |
24 |     uri!(has_one: 1, 23); //~ ERROR expects 1 parameter but 2
   |                   ^^^^^
   |
   = note: expected parameter: id: i32

error: `has_one` route uri expects 1 parameter but 0 were supplied
  --> $DIR/typed-uris-bad-params.rs:22:10
   |
22 |     uri!(has_one); //~ ERROR expects 1 parameter but 0
   |          ^^^^^^^
   |
   = note: expected parameter: id: i32

error: aborting due to 21 previous errors

